marshaler.go

```diff
@@ -165,28 +165,37 @@ func (ctx *encoderCtx) isRoot() bool {
 	return len(ctx.parentKey) == 0 && !ctx.hasKey
 }
 
-var errUnsupportedValue = errors.New("unsupported encode value kind")
-var errTextMarshalerCannotBeAtRoot = errors.New("type implementing TextMarshaler cannot be at root")
+var (
+	errInterfaceNotTime            = errors.New("timeType but interface not time")
+	errUnsupportedValue            = errors.New("unsupported encode value kind")
+	errTextMarshalerCannotBeAtRoot = errors.New("type implementing TextMarshaler cannot be at root")
+)
 
-//nolint:cyclop
+//nolint:cyclop,funlen
 func (enc *Encoder) encode(b []byte, ctx encoderCtx, v reflect.Value) ([]byte, error) {
-	//nolint:gocritic,godox
-
 	if v.Type() == timeType {
-		i := v.Interface().(time.Time)
+		i, ok := v.Interface().(time.Time)
+		if !ok {
+			return nil, errInterfaceNotTime
+		}
+
```

Comment:
very unlikely to happen (maybe even impossible to have a timeType not castable into a time, but... well, in case the code around is changing, this might catch something one day.

Reply:
I ll bench it and come back with that. I feel it's clearer and simpler to just have the cast with the ok, but let's see, I feel weak with my assumptions now :P

Reply:
```golang
var timeType = reflect.TypeOf(time.Time{})

func v1(v reflect.Value) time.Time {
	if v.Type() == timeType {
		return v.Interface().(time.Time)
	}

	return time.Time{}
}

func v2(v reflect.Value) time.Time {
	i, ok := v.Interface().(time.Time)
	if ok {
		return i
	}

	return time.Time{}
}
```
```golang
func BenchmarkV1(b *testing.B) {
	x := []reflect.Value{reflect.ValueOf(0), reflect.ValueOf(time.Now())}

	b.ReportAllocs()
	b.ResetTimer()

	// run the Fib function b.N times
	for n := 0; n < b.N; n++ {
		v1(x[n%2])
	}
}

func BenchmarkV2(b *testing.B) {
	x := []reflect.Value{reflect.ValueOf(0), reflect.ValueOf(time.Now())}

	b.ReportAllocs()
	b.ResetTimer()

	// run the Fib function b.N times
	for n := 0; n < b.N; n++ {
		v2(x[n%2])
	}
}
```

```
name  time/op
V1-8  9.36ns ± 1%
V2-8  6.42ns ± 2%

name  alloc/op
V1-8   0.00B     
V2-8   0.00B
```

so it seems better to simply try to cast and check if ok. It's also clearer imho.

Reply:
sounds good to me!